core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -139,43 +139,39 @@ static boolean execute(SearchContext searchContext, final IndexSearcher searcher
             assert query == searcher.rewrite(query); // already rewritten
 
             final ScrollContext scrollContext = searchContext.scrollContext();
-            if (scrollContext != null) {
-                if (scrollContext.totalHits == -1) {
-                    // first round
-                    assert scrollContext.lastEmittedDoc == null;
-                    // there is not much that we can optimize here since we want to collect all
-                    // documents in order to get the total number of hits
-
-                } else {
-                    final ScoreDoc after = scrollContext.lastEmittedDoc;
-                    if (returnsDocsInOrder(query, searchContext.sort())) {
-                        // now this gets interesting: since we sort in index-order, we can directly
-                        // skip to the desired doc
-                        if (after != null) {
-                            BooleanQuery bq = new BooleanQuery.Builder()
-                                .add(query, BooleanClause.Occur.MUST)
-                                .add(new MinDocQuery(after.doc + 1), BooleanClause.Occur.FILTER)
-                                .build();
-                            query = bq;
-                        }
-                        // ... and stop collecting after ${size} matches
-                        searchContext.terminateAfter(searchContext.size());
-                        searchContext.trackTotalHits(false);
-                    } else if (canEarlyTerminate(indexSort, searchContext)) {
-                        // now this gets interesting: since the index sort matches the search sort, we can directly
-                        // skip to the desired doc
-                        if (after != null) {
-                            BooleanQuery bq = new BooleanQuery.Builder()
-                                .add(query, BooleanClause.Occur.MUST)
-                                .add(new SearchAfterSortedDocQuery(indexSort, (FieldDoc) after), BooleanClause.Occur.FILTER)
-                                .build();
-                            query = bq;
-                        }
-                        searchContext.trackTotalHits(false);
-                    }
+            final boolean isScrollContext = scrollContext != null && scrollContext.totalHits != -1;
+            final ScoreDoc after = isScrollContext ? scrollContext.lastEmittedDoc : searchContext.searchAfter();
+            if (returnsDocsInOrder(query, searchContext.sort())) {
+                // now this gets interesting: since we sort in index-order, we can directly
+                // skip to the desired doc
+                if (after != null &&
+                        (isScrollContext || searchContext.trackTotalHits() == false)) {
```

Comment:
in the non-scroll case, we should also check that there are no aggs?

Reply:
My first goal in this PR is to be able to filter documents in top_hits **and** aggs in order to be able to use `search_after` in sorted streaming aggregations so this is done in purpose. I can add a note in the docs regarding this behavior ?

Reply:
OK, in that case I think a query (like your other PR did) is more appropriate since there is a general contract that `query` filters what aggregations can see but `search_after` doesn't. However I'd like to understand better how things will look like in the end since I'm a bit on the fence about having features that only work with sorted indices.